### PR TITLE
Add max length for the slug

### PR DESCRIPTION
**Pull request by Arsalanliaqat**

### CODE REVIEW

File: lib/slugger.ts [hunk 1]
```diff
@@ -62,13 +62,22 @@ export interface SluggerInitOptions<D extends Document> {
...
+   */
+  maxLength?: number;
 }
 
 export class SluggerOptions<D extends Document> {
   readonly slugPath: string;
   readonly generator: GeneratorFunction<D>;
   readonly index: string;
   readonly maxAttempts?: number;
+  readonly maxLength?: number;
```
Comment (qqilihq) [inline lib/slugger.ts:80]: Add a validation that `maxLength` has reasonable value (i.e. > 0)

**Review by qqilihq — CHANGES_REQUESTED** [review]

**Comment by qqilihq** [comment]
Thank you, Arslan. Beside the comments above, could you please add test cases which cover the new maxLength functionality? As far as I can see there were some in the older branch which I once did, probably you can just copy these.

File: lib/sluggerUtils.ts [hunk 2]
```diff
@@ -59,16 +59,22 @@ export async function saveSlugWithRetries<D extends Document>(
...
-  return (doc, attempt) => {
+  return (doc: Document, attempt: number, maxLength?: number) => {
     const values = ([] as string[]).concat(paths).map(path => doc.get(path) as string);
     if (attempt > 0) {
       values.push(`${attempt + 1}`);
     }
     // replace underscore with hyphen
-    return limax(values.join('-'), { custom: { _: '-' } });
+    const slug = limax(values.join('-'), { custom: { _: '-' } });
+    return maxLength && maxLength > 0 ? trimSlug(slug, maxLength) : slug;
```
[the remark this answers is not in the document]
Comment (Arsalanliaqat) [reply lib/sluggerUtils.ts:69]: What do you suggest here? Does it make sense to trim from the middle section of the string?

`this-is-another-valid-example-of-a-very-very-long-slug-1` to something like this `this-is-another-very-long-slug-1`.

[the remark this answers is not in the document]
Comment (qqilihq) [reply lib/sluggerUtils.ts:69]: Let's just trim the original string, and then add the running index if needed. Eg. slug length is 30 characters:

`this-is-another-valid-example-of-a-very-very-long-slug-1` -> `this-is-another-valid-examp-1`

File: readme.md [hunk 3]
```diff
@@ -63,7 +63,8 @@ const sluggerOptions = new slugger.SluggerOptions({
   // the property path which stores the slug value
   slugPath: 'slug',
   // specify the properties which will be used for generating the slug
-  generateFrom: [ 'firstname', 'lastname' ],
+  generateFrom: ['firstname', 'lastname'],
+  maxLength: 0,
```
Comment (qqilihq) [inline readme.md:67]: Give a valid example here

**Comment by qqilihq** [comment]
Hey Arslan - when I currently run the tests locally, one of them fails. Does it work for you? Could you look into this?

```
  ● slugger › saving to database › generating duplicate slugs within one sequence › throws when same slugs are generated within one save cycle

    expect(received).toBeInstanceOf(expected)

    Expected constructor: SluggerError
    Received constructor: ReferenceError

      487 |           fail();
      488 |         } catch (e) {
    > 489 |           expect(e).toBeInstanceOf(slugger.SluggerError);
          |                     ^
      490 |           expect((e as slugger.SluggerError).message).toEqual("Already attempted slug 'john' before. Giving up.");
      491 |         }
      492 |       });
```

**Comment by qqilihq** [comment]
One more thing: As far as I read the code, you currently take the `maxLength` either from the slugger options or directly from the schema. That's cool, but could you explicitly document this in the readme?

**Comment by Arsalanliaqat** [comment]
> Hey Arslan - when I currently run the tests locally, one of them fails. Does it work for you? Could you look into this?
> 
> ```
>   ● slugger › saving to database › generating duplicate slugs within one sequence › throws when same slugs are generated within one save cycle
> 
>     expect(received).toBeInstanceOf(expected)
> 
>     Expected constructor: SluggerError
>     Received constructor: ReferenceError
> 
>       487 |           fail();
>       488 |         } catch (e) {
>     > 489 |           expect(e).toBeInstanceOf(slugger.SluggerError);
>           |                     ^
>       490 |           expect((e as slugger.SluggerError).message).toEqual("Already attempted slug 'john' before. Giving up.");
>       491 |         }
>       492 |       });
> ```

This is strange; It works fine for me.

image

**Pull request merged**